src/runtime/program/compute_budget/lib.zig

```diff
@@ -109,8 +111,169 @@ const ComputeBudgetInstruction = union(enum(u32)) {
 
     /// Set a specific transaction-wide account data size limit, in bytes, is allowed to load.
     set_loaded_accounts_data_size_limit: u32,
+
+    pub const Tag = @typeInfo(ComputeBudgetInstruction).@"union".tag_type.?;
+
+    const borsh = BorshStaticTaggedUnionHelper(ComputeBudgetInstruction);
+
+    pub const BorshDecodeError = error{
+        /// The raw data did not match the expected representation.
+        DecodeFailed,
+        /// `src` was too short.
+        TruncatedSrc,
+    };
+
+    /// Decodes the encoded ComputeBudgetInstruction, and the number of bytes it occupied in `src`.
+    pub fn borshDecodeSlice(
+        src: []const u8,
+    ) BorshDecodeError!struct { ComputeBudgetInstruction, borsh.EncodedSizeInt } {
+        if (src.len == 0) return error.TruncatedSrc;
+        const tag = borsh.tagDecode(src[0]) orelse return error.DecodeFailed;
+        const pl_size: borsh.EncodedSizeInt = borsh.payloadEncodedSize(tag);
+        const pl_bytes = src[1..];
+        if (pl_bytes.len < pl_size) return error.TruncatedSrc;
+        const cbi = borsh.decodeFromTagAndPayload(tag, pl_bytes[0..pl_size]) orelse
+            return error.DecodeFailed;
+        return .{ cbi, @sizeOf(u8) + pl_size };
+    }
 };
 
+// TODO: put this somewhere logical, maybe create a more general abstraction?
+fn BorshStaticTaggedUnionHelper(comptime Tagged: type) type {
+    const Tag, const has_payloads = switch (@typeInfo(Tagged)) {
+        .@"union" => |union_info| .{ union_info.tag_type.?, true },
+        .@"enum" => .{ Tagged, false },
+        else => @compileError("Unsupported type " ++ @typeName(Tagged)),
+    };
+    if (@sizeOf(Tag) > @sizeOf(u8)) @compileError("Tag type is too big");
+
+    const pl_size_min, const pl_size_max = blk: {
+        var min: usize = @sizeOf(Tagged);
+        var max: usize = 0;
```

Comment:
If there are no fields, the minimum is larger than the maximum. Why don't min and max both start at the same value? It seems like maybe they should both start at 0, since the tag is given 1 byte and you account for that separately.

Reply:
min can't start at 0, because then it would always be smaller than any of the fields.

Reply:
I think `min` should start at `maxInt`. It will be too small if the tag is smaller than all of the fields.

I found this code pretty confusing at first. I'm looking for ways to simplify it and make it less esoteric, so anyone can quickly look at it and understand it easily.

1. `min` is unnecessary because you can simply use 0 as the `from` value in IntFittingRange and it will return the same type.
2. EncodedSizeInt and EncodedPayloadSizeInt could be replaced with `usize` for clarity.

You could probably eliminate all the pub consts except for `ENCODED_SIZE_MAX`.

Are there reasons for all these extra pieces? I think the code would be easier to understand without them.

Reply:
723dcc527a5b26eddd06cd3c56a609c85ef28802